### PR TITLE
remove componentwise comparison of complex values in torch.testing.assert_close

### LINKED ISSUE #61906: RESOLVED: Retire component-wise closeness tests for complex tensors
Closeness checks in our the test suite, i.e. `TestCase.assertEqual`, are performed component wise meaning closeness is computed for the real and imaginary components separately.

https://github.com/pytorch/pytorch/blob/39ce29efe05dc09fe9de6919715548476e4ab7b6/torch/testing/_core.py#L137-L164

The recently introduced `torch.testing.assert_close` inherited the behavior:

https://github.com/pytorch/pytorch/blob/39ce29efe05dc09fe9de6919715548476e4ab7b6/torch/testing/_asserts.py#L72-L88

The componentwise checking was introduced in #34258 but no public explanation was provided on the reason behind it.

Pros:

- Since `complex64` is comprised of two `float32`, we can reuse the testing tolerances. Of course the same is true for `complex128` and `complex32`.

Cons:

- `torch.testing.assert_close` and `torch.isclose` diverge for complex tensors:

    ```python
    >>> rtol, atol = 0.1, 0.3
    >>> expected = torch.tensor(1 + 1j)
    >>> actual1 = torch.tensor(1 + 1.43j)
    >>> torch.testing.assert_close(actual1, expected, rtol=rtol, atol=atol)
    AssertionError: Scalars are not close!
    >>> torch.isclose(actual1, expected, rtol=rtol, atol=atol)
    tensor(True)
    >>> actual2 = torch.tensor(0.6 + 1.3j)
    >>> torch.testing.assert_close(actual2, expected, rtol=rtol, atol=atol)
    >>> torch.isclose(actual2, expected, rtol=rtol, atol=atol)
    tensor(False)
    ```

    `torch.isclose` follows the implementation of `numpy` and partially `cmath` by comparing the absolute values of the difference of the two values. `torch.testing.assert_close` on the other hand compares the absolute values of the difference of the individual components. To be exact, `torch.isclose` as well as `numpy` evaluate the closeness inequality `abs(a - b) <= atol + rtol * abs(b)` the same way for complex as for real numbers. `torch.testing.assert_close` evaluates it twice for the real and imaginary component and checks the logical and of the results.
- The implementation is more messy and introduces more edgecases:

    - Is `complex(inf, 0)` close to `complex(inf, 1)`?
    - Is `complex(0, NaN)` close to `complex(NaN, 0)`?

    The second case led to `equal_nan="relaxed"` in `TestCase.assertEqual` which accepts these comparisons while they will be rejected for `equal_nan=True`. `torch.isclose` currently bails out for complex values and `equal_nan=True`

    https://github.com/pytorch/pytorch/blob/a26a9f8b75f4e8c9c7c34907a099b3a1d41f8a2c/aten/src/ATen/native/TensorCompare.cpp#L101-L102

## Regions of closeness

By checking the components individually we are forming a rectangle region of closeness around the expected value. Despite the divergence on the tolerance definition, `torch.isclose`, `numpy`, and `cmath` form a circle around the expected values. Thus is impossible to convert the old tolerances to achieve the same behavior if we go for this change.

## RFC

Since in most cases the specific values of the tolerances hardly matter, IMHO the cons severly outweigh the pros. Thus, the current plan is to retire the componentwise checking and merge the complex behavior of our internal and public testing functions with `torch.isclose` and in turn with `numpy`. If you disagree with that plan, please let us know in this issue.

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @VitalyFedyunin @walterddr

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63841
* #63572

Closes #61906.

cc @ezyang @gchanan

Differential Revision: [D30633526](https://our.internmc.facebook.com/intern/diff/D30633526)

### CODE REVIEW

File: test/test_testing.py [hunk 1]
```diff
@@ -1180,30 +1194,6 @@ def test_mapping_mismatching_values_msg(self):
...
-                fn(equal_nan=True)
-
-    def test_mismatching_nan_with_equal_nan_relaxed(self):
-        actual = torch.tensor(complex(1, float("NaN")))
-        expected = torch.tensor(complex(float("NaN"), 1))
-
-        for fn in assert_close_with_inputs(actual, expected):
-            fn(equal_nan="relaxed")
-
-    def test_matching_conjugate_bit(self):
```
Comment (mruberry) [inline test/test_testing.py:1199]: Don't we still need this test?

File: torch/testing/_asserts.py [hunk 2]
```diff
@@ -429,11 +383,18 @@ def append_difference(msg: str, *, type: str, difference: float, index: Tuple[in
 
     return msg.strip()
 
+def _get_XXX_dtype(dtype: torch.dtype) -> torch.dtype:
```
Comment (mruberry) [inline torch/testing/_asserts.py:386]: _get_comparison_dtype?

**Review by mruberry — COMMENTED** [review]
Cool simplification but maybe a little too aggressive with the conjugate bit test? I also have a name suggestion (naming things is hard but XXX seems easy to improve on)

File: torch/testing/_asserts.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_asserts.py:386]: Add a comment for what this function does and how it's intended to be used

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30633526).

File: torch/testing/_asserts.py (hunk 2 above)
Comment (pmeier) [reply torch/testing/_asserts.py:386]: Oops :innocent:

File: test/test_testing.py (hunk 1 above)
Comment (pmeier) [reply test/test_testing.py:1199]: #56058 removed the parts that this test was designed for. So yes, we can keep it, but I don't think we need to. cc @anjali411

Comment (mruberry) [reply test/test_testing.py:1199]: How about we keep it for now (just in case) but add a TODO: over it that we should review later?

**Review by mruberry — APPROVED** [review]
Cool!

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30633526).

**Pull request closed without merging**